synapseclient/producer_consumer_download_threads.py

```diff
@@ -0,0 +1,282 @@
+    """
+    The producer threads that make the GET request and obtain the data for a download chunk
+    """
+    def __init__(self, client, request, range_queue, data_queue):
+        Thread.__init__(self)
+        self.setDaemon(True)
+        self.client = client
+        self.request = request
+        self.range_queue = range_queue
+        self.data_queue = data_queue
+        self.session = requests.Session()
+        retry = Retry(connect=MAX_RETRIES, backoff_factor=BACK_OFF_FACTOR)
+        adapter = HTTPAdapter(max_retries=retry)
+        self.session.mount('http://', adapter)
+        self.session.mount('https://', adapter)
+
+    def run(self):
+        while True:
+            # Get the work from the chunk generator
+            start, end, file_name, url, path = self.range_queue.get()
+            try:
+                # download chunk
+                # specify the start and end of this chunk
+                headers = {'Range': 'bytes=%d-%d' % (start, end)}
+
+                if not _url_is_valid(url):
+                    _, pre_signed_url_new = _get_pre_signed_batch_request_json(self.client, self.request)
+                    url = pre_signed_url_new
+                # request the chunk and capture the response
+                response = self.session.get(url, headers=headers, stream=True)
+                # try request until successful or out of retries
+                try_counter = 0
+                while response.status_code != PARTIAL_CONTENT_CODE and try_counter < MAX_RETRIES:
+                    try_counter += 1
+                    if not _url_is_valid(url):
+                        _, pre_signed_url_new = _get_pre_signed_batch_request_json(self.client, self.request)
+                        url = pre_signed_url_new
+                    response = self.session.get(url, headers=headers, stream=True)
+
+                if response.status_code == PARTIAL_CONTENT_CODE:
+                    self.data_queue.put((start, file_name, path, response.content))
+            finally:
+                self.range_queue.task_done()
+
+
+class ConsumerDownloadThread(Thread):
+    """
+    The worker threads that write download chunks to file
+    """
+    def __init__(self, data_queue):
+        Thread.__init__(self)
+        self.setDaemon(True)
+        self.data_queue = data_queue
+
+    def run(self):
+        while True:
+            # Get the chunk data from the data queue
+            start, file_name, path, data = self.data_queue.get()
+            try:
+                # write data to file
+                with open(path, "r+b") as file_write:
+                    global transferred
+                    global t_first_write
+                    global previously_transferred
+                    if transferred == 0:
+                        t_first_write = time.time()
+                    file_write.seek(start)
+                    file_write.write(data)
+
+                    transferred = len(data) + previously_transferred
```

Comment:
Look into removing one of these, i.e. previously transferred.

Reply:
Got rid of previously_transferred, simply incremented transferred: transferred += len(data)